src/TabNine.ts

```diff
@@ -1,300 +1,124 @@
-        return { enabled_features: [] };
-      }
-      return result;
-    } catch (error) {
-      console.error(error);
-      return { enabled_features: [] };
-    }
+  constructor(private binaryRun: BinaryRun) {
+    this.restartChild();
   }
 
-  private requestUnlocked(
-    version: string,
-    any_request: any,
-    timeout = 1000
-  ): Promise<any> {
-    any_request = {
-      version: version,
-      request: any_request,
-    };
+  public async request(request: any, timeout = 1000): Promise<any> {
+    console.log("Sending request to binary: ", request);
 
-    const unregisterFunctions = [];
-
-    const request = JSON.stringify(any_request) + "\n";
+    const release = await this.mutex.acquire();
 
-    let response = new Promise<any>((resolve, reject) => {
-      try {
-        if (!this.isChildAlive()) {
-          this.restartChild();
-        }
-        if (!this.isChildAlive()) {
-          reject(new Error("TabNine process is dead."));
-        }
-        const onResponse: (input: any) => void = (response) => {
-          let any_response: any = JSON.parse(response.toString());
-          resolve(any_response);
-        };
-        this.rl.once("line", onResponse);
+    try {
+      if (this.isBinaryDead()) {
+        this.restartChild();
 
-        unregisterFunctions.push(() =>
-          this.rl.removeListener("line", onResponse)
-        );
-        this.proc.stdin.write(request, "utf8");
-      } catch (e) {
-        console.log(`Error interacting with TabNine: ${e}`);
-        reject(e);
+        throw new Error("TabNine process is dead.");
       }
-    });
 
-    let timer = new Promise((_resolve, reject) => {
-      let timer = setTimeout(() => reject("request timed out"), timeout);
-
-      unregisterFunctions.push(() => clearTimeout(timer));
-    });
-
-    let procExit = new Promise((_resolve, reject) => {
-      const onClose = () => reject("Child process exited");
-      this.proc.once("exit", onClose);
+      this.proc.stdin.write(
+        JSON.stringify({
+          version: API_VERSION,
+          request: request,
+        }) + "\n",
+        "utf8"
+      );
 
-      unregisterFunctions.push(() => this.proc.removeListener("exit", onClose));
-    });
+      const result = await this.readLineWithLimit(timeout);
 
-    const unregister = () => {
-      unregisterFunctions.forEach((f) => f());
-    };
+      console.log("Received result from binary: ", result);
+      this.consecutiveRestarts = 0;
 
-    return Promise.race([response, timer, procExit]).then(
-      (value) => {
-        unregister();
-        return value;
-      },
-      (err) => {
-        unregister();
-        throw err;
-      }
-    );
+      return JSON.parse(result.toString());
+    } finally {
+      console.log("Binary request failed.");
+      release();
+    }
   }
 
-  private isChildAlive(): boolean {
-    return this.proc && !this.childDead;
-  }
+  private readLineWithLimit(timeout: number): Promise<any> {
```

Comment:
This makes the whole old mechanism unneeded. I just wrapped it with an exception, and it just works.